N°2651 - Removal of "Test" dirs within dependencies handled using composer

bruno-ds opened the pull request:
Note: 

since it was the problem with the previous solution, the `iTopComposer::ListAllowedTestDir()`
MUST be reread, for now it only contains two entries : 

			APPROOT.'lib/twig/twig/src/Node/Expression/Test/',
			APPROOT.'lib/twig/twig/lib/Twig/Node/Expression/Test/',

Commit:
N°2651 - Removal of "Test" dirs within dependencies handled using composer

piRGoif commented on sources/application/composer/iTopComposer.php:
```diff
@@ -0,0 +1,100 @@
+			APPROOT.'lib/symfony/var-dumper/Tests/',
+			APPROOT.'lib/symfony/web-profiler-bundle/Tests/',
+			APPROOT.'lib/symfony/yaml/Tests/',
+			APPROOT.'lib/symfony/debug/Resources/ext/tests/',
```
For the following :

'lib/symfony/debug/Resources/ext/tests/'
'lib/symfony/framework-bundle/Tests/Fixtures/TestBundle/FooBundle/Controller/Test/'

Might not be a good idea to remove... are we sure they really contains tests only ?

bruno-ds replied:
after analysing this : 
 - `lib/symfony/debug/Resources/ext` is C code (a php extention), so it is out of scope and the test dir can be excluded.

 - `lib/symfony/framework-bundle/Tests` is a test dir, so it subdir can be excluded

Molkobain commented:
Seems all good to me. Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place? This folder is meant mostly for the GUI of the admin. console. Also, shouldn't we put a namespace to it if it's in the /sources folder like all the other files ?

bruno-ds commented:
I admit that I struggled to find where to put it, and wasn't happy with my choice.
Even more now that you point the GUI related semantic of this for.
Feel free to move it :)

For the namespace, has said many sages during hard time: whoops!

piRGoif commented:
> Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place?

I agree too !
Why not simply /lib or /lib/bin ?

piRGoif commented:
> 
> > Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place?
>
> Why not simply /lib or /lib/bin ?

Discussed with Bruno : not a good idea, as this dir content is handled by Composer !
Also, the goal was to have a specific namespace in iTopComposer.php, and so putting it in /application made it working natively with our autoloader.

piRGoif committed:
N°2651 Remove tests from lib : browse dirs using SPL classes instead of GLOB

piRGoif commented:
@bruno-ds I've juste pushed (5187834) a modification for the \iTopComposer::ListAllTestDir method : use Spl classes instead of multiple glob patterns.
Results are the same now (I fixed some problems on windows with DIRECTORY_SEPARATOR)

bruno-ds commented on sources/application/composer/iTopComposer.php:
```diff
@@ -0,0 +1,126 @@
+		$oDirectoryIterator = new RecursiveDirectoryIterator($sPath);
+		$iterator = new RecursiveIteratorIterator(
+			$oDirectoryIterator,
+			RecursiveIteratorIterator::CHILD_FIRST);
```
https://www.php.net/manual/fr/class.filesystemiterator.php#filesystemiterator.constants

I propose : 
FilesystemIterator::SKIP_DOTS
FilesystemIterator::UNIX_PATHS

piRGoif replied:
Yeah, seems perfect ! Can you add it as a code suggestion ? (https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request see point 6)

Molkobain commented:
I'm still troubled by this file being in the /sources/application folder as it is really not part of the admin. console GUI... 😕

As this is more of an internal tool or something, would a /bin or /.make be suitable?

piRGoif commented:
> As this is more of an internal tool or something, would a /bin or /.make be suitable?

Indeed, I would recommend renaming .make to .bin, and moving iTopComposer.php file there ?

Commit:
N°1627 - Ticket ref sometimes duplicate
:bug: KeyValueStore do no more create multiples keys if the current user have no read access

odain-cbd committed:
Merge branch 'develop' into feature/composer_without_tests

odain-cbd committed:
added namespace + mv iTopComposer + optimize FileIterator

odain-cbd commented:
Pull request already merged manually.

piRGoif commented:
Updates tags & milestone :)
Are you ok to delete branch ?

odain-cbd commented:
Thanks for the reminder. we planned to do it with Bruno and forgot...
